docs/patterns/outbox.md

```diff
@@ -0,0 +1,141 @@
+- Requires that the primary resource supports optimistic concurrency control
+- Requires storing information about all previously processed messages. In principle that information should never be removed but in practice often age-based approach is used to prevent data from growing indefinitely. 
+
+## Issues and considerations
+
+Consider the following points when implementing the Outbox pattern:
+
+- In order to take advantage of the pattern, each message needs to have a unique identifier. Many messaging technologies like Azure Service Bus provide this identifier out-of-the-box, but in some of them, it is not set automatically.
+- Consider how long to store information about processed messages. The likelihood of message being duplicated diminishes with time significantly. In practice value of 1 or 2 weeks is often used as maximum age to keep the deduplication data.
+- Depending on the technology used for the primary resource, it might or might not be possible to use native features to remove expired dedupliction data. For example, Cosmos DB has a built in time-to-live mechanism that can be used to ensure message processing information is removed after configured time.
+- The clock for removing deduplication data can only be started after the `CleanUpSideEffects` operation succeeds. Otherwise there is a risk of losing the side effects if a message is stuck for extended period of time, e.g. falsly identified as a _poison message_ and sent to a _dead-letter queue_.
+- Side effects other than sending or publishing messages are supported provided the likelihood of failure to apply them is negligible, e.g. creating or updating a document in a key-value store.
+
+### Side effects application
+
+The design of the side effects application is a major consideration when implementing the outbox pattern. There are three major approaches to the problem.
+
+#### Background worker
+
+In this approach an external process periodically polls the data store where the side effects are persisted. The advantage of this approach is the fact that it can also be used at the boundary of the system e.g. when processing incoming HTTP requests and sending out messagess. There are, however, significant disadvantages of this approach. First, it relies on querying the data store for all outstanding side effects. This query pattern might be expensive or even impossible in certain partitioned data stores (e.g. Cosmos DB). The second issue is the fact that polling implies the code is executed periodically. Such requirement makes deployment in serverless environments (e.g. Azure Functions) problematic as the worker needs to use a dedicated (timer) trigger. Finally, the polling mechanism has poor scalability characteristics. Higher side effect processing throughput cannot be simply achieved by deployment multiple background workers as they would compete to process the same items. 
```

Comment:
```suggestion
In this approach, an external process periodically polls the data store where the side effects are persisted. The advantage of this approach is the fact that it can also be used at the boundary of the system e.g. when processing incoming HTTP requests and sending out messages. There are, however, significant disadvantages. 

First, it relies on querying the data store for all outstanding side effects. This query pattern might be expensive or even impossible in certain partitioned data stores (e.g. Cosmos DB). Secondly, polling requires the code to be executed periodically (at intervals). Such a requirement makes deployment in serverless environments (e.g. Azure Functions) problematic as the worker needs a dedicated (timer) trigger. Finally, the polling mechanism has poor scalability characteristics. Higher side effect processing throughput cannot be simply achieved by deploying multiple background workers as they would compete to process the same items. 
```